### ISSUE TITLE
Add coupon information to data passed to canMakePayment

### ISSUE THREAD

**Issue by ralucaStan** [woocommerce/woocommerce-gutenberg-products-block#4670]
The Conditional Shipping and Payments extension needs to have access to the coupon code data present in the cart.
It should have this info available after each change in the checkout block in order to determine if a payment method can be used. 

Below is the date needed by the extension.

![Screenshot 2021-09-01 at 18.08.15.png](https://images.zenhubusercontent.com/6006f97ad883b38577590742/a2bf459f-b885-4564-8d91-56752ec0710a)

This can be achieved if we pass coupon data to this [canMakePayment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/ca3b3f0fda1582c919b13e538cccca9f6733ab3d/docs/extensibility/payment-method-integration.md#canmakepayment-required) callback.

**Comment by opr** [comment]
If we pass the `cart` to the filter then this will be available.

```
cartData: {
      coupons: [
        {
          code: '10off',
          discount_type: 'percent',
          totals: {
            total_discount: '2070',
            total_discount_tax: '0',
            currency_code: 'GBP',
            currency_symbol: '£',
            currency_minor_unit: 2,
            currency_decimal_separator: '.',
            currency_thousand_separator: ',',
            currency_prefix: '£',
            currency_suffix: ''
          }
        }
      ],
      ...
}
```

**Comment by ralucaStan** [comment]
There is logic in the project for canPayArgument, the argument that is passed to the [canMakePayment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/ca3b3f0fda1582c919b13e538cccca9f6733ab3d/assets/js/base/context/providers/cart-checkout/payment-methods/use-payment-method-registration.ts#L69), so we'd want to keep it as specific as possible because each time one of the values it contains changes, then the canMakePayment logic is rerun. So we would want to limit the reruns.
Here is what I had in mind:

```
canMakePayment( {
    // add just coupons entry
    cartCoupons: ...,
    cartTotals: CartTotals,
    cartNeedsShipping: boolean,
    shippingAddress: CartShippingAddress,
    billingData: BillingData,
    selectedShippingMethods: string[],
    paymentRequirements: string[],
} )
```

**Comment by opr** [comment]
I was experimenting with adding `cart` to this, I don't think it's a bad thing or performance degradation. I say this because I think the ways in which the `cart` can change would cause `cartTotals` to change anyway, right?
- Remove item from cart
- Change quantity of item in cart
- Change shipping method
- Add coupon

**Comment by ralucaStan** [comment]
closed by #4776
